### PR TITLE
Deduplicate ProjectReferences on VS nomination

### LINKED ISSUE #11639: [Bug]: NU1004 in Visual Studio, but not command line (lock files in locked mode)
### NuGet Product Used

Visual Studio Package Management UI, Visual Studio Package Manager Console

### Product Version

VS16.11

### Worked before?

_No response_

### Impact

It's more difficult to complete my work

### Repro Steps & Context

An internal customer reported this to me in a huge, extremely complex, closed source repo. Given the complexity, I'm not 100% sure how to reproduce it, but the outcome is that in Visual Studio, project-system was sending NuGet a nomination where the project had two ProjectReferences to the same project. One had path `..\some\reference.csproj`, whereas the other had path `d:\path\to\some\reference.csproj`. I guess this is why project-system failed to de-duplicate before sending to NuGet.

If the Directory.Build.props (or in this case some organization mandated SDK/build system) uses `<RestoreLockedMode>true`, then VS will fail with [NU1004](https://docs.microsoft.com/nuget/reference/errors-and-warnings/nu1004). The error message is super difficult to understand/determine what the difference is, but if you spend enough time, you can figure out that some of project references are duplicated. Figuring out why the project references are duplicated is even more difficult, and maybe is not possible by customers, even with a binlog. I had to debug NuGet's VS extension to figure it out.

### Verbose Logs

_No response_

### PR DESCRIPTION
**Pull request by zivkan**
## Bug

Fixes: https://github.com/NuGet/Home/issues/11639

Regression? probably no

## Description

I'm not sure how exactly, but a project was nominated with a ProjectReference to `..\some\project.csproj` and another ProjectReference to `d:\path\to\some\project.csproj`. When using lock files and locked mode, this caused NuGet to expect the project to be in the lock file multiple times, which is was not (lock file generated from the command line?  or maybe writing the lock file correctly de-dupes, but the compare did not?).

Anyway, added a little logic to deduplicate the project references on nomination.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

**Review by nkolev92 — APPROVED** [review]
Isn't there a project system bug here as well?

**Comment by zivkan** [comment]
Maybe, but unless I can find a simple repro, I don't know if creating a bug where repro steps include "clone this internal private repo, and spend 4 hours setting it up" is going to get much traction.

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/VSNominationUtilities.cs [hunk 1]
```diff
@@ -187,7 +187,8 @@ internal static NuGetFramework GetTargetFramework(IVsProjectProperties propertie
                     targetFrameworkInfo.ProjectReferences
                         .Cast<IVsReferenceItem>()
                         .Where(IsReferenceOutputAssemblyTrueOrEmpty)
-                        .Select(item => ToProjectRestoreReference(item, projectDirectory)));
+                        .Select(item => ToProjectRestoreReference(item, projectDirectory))
+                        .Distinct(ProjectRestoreReferenceComparer.Default));
```
Comment (kartheekp-ms) [inline src/NuGet.Clients/NuGet.SolutionRestoreManager/VSNominationUtilities.cs:191]: Given that `ProjectRestoreReference` implements `IEquatable`, As per second example in https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.distinct?view=net-6.0 directly invoking `.Distinct()` should work without passing custom comparer. It is possible that I may be missing an important scenario.

https://github.com/NuGet/NuGet.Client/blob/71424bcf70087a60abc6ec6b86035bd572400464/src/NuGet.Core/NuGet.ProjectModel/ProjectRestoreReference.cs#L13

Comment (zivkan) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/VSNominationUtilities.cs:191]: ProjectRestoreReference's implementation compares not only project path/unique name, but also IncludeAssets, ExcludeAssets and PrivateAssets.  If for some reason these are different, then we'll have the same problem with there being duplicate references to the same project.

**Review by jeffkl — APPROVED** [review]

**Review by erdembayar — APPROVED** [review]

**Comment by zivkan** [comment]
I couldn't repro the locked mode restore error of partner team bug report I was debugging, but I reproed something similar enough and reported to project system team: https://github.com/dotnet/project-system/issues/7952

**Pull request merged**